### DISCUSSION TITLE
[Question] Conditionally set file extension?

### DISCUSSION THREAD

**Discussion by Ajaymamtora** [twpayne/chezmoi#3963] [Q&A] [answered]
Is there a way to have a single file conditionally have a file extension? At the minute I have 2 separate files for the same thing for windows/linux:
![image](https://github.com/user-attachments/assets/60db12f3-0ff8-4d64-ab67-39acbd27c1b4)

Where the windows one is wrapped in:

```go
{{ if eq .chezmoi.os "windows" -}}
```

and the linux one is:

```go
{{ if ne .chezmoi.os "windows" -}}
```

and it would be good to condense these into 1 file

**Comment by halostatue** [comment, marked as answer]
No, there is not.

There are a few options available to you:

- Scripts can be ignored in `.chezmoiignore.tmpl` just as they do to other files (see [twpayne/dotfiles/home/.chezmoiignore.tmpl](https://github.com/twpayne/dotfiles/blob/7d37dee723f74ede053d68c5104b58a536c8b8a5/home/.chezmoiignore.tmpl#L4-L19) (this prevents the Windows one from running on Linux, but does not unify the scripts).
- A version of bash can be installed on Windows (this can potentially unify the scripts).
- A version of PowerShell can be installed on Linux (this can potentially unify the scripts).
- Your script can be written in a different common language (Ruby, Python, Tcl, whatever) with the appropriate interpreter installed (unifying the scripts).
- You can *always* call the file `.ps1.tmpl` but do something like:

  ```
  {{- if eq .chezmoi.os "windows" -}}
  # PowerShell script
  {{- else -}}
  #! /usr/bin/env bash
  # bash script
  {{ end -}}
  ```

  This works because Linux & Unix systems do not meaningfully use extensions (they may mean something to the desktop environment, but not to the shell) and the shebang (`#! /usr/bin/env bash`) is all that matters. This will work, but maintenance will be somewhat harder because your editor will not think that your bash script is valid (and the template code will also be somewhat uglier). This can be further mitigated with judicious use of [`include`](https://www.chezmoi.io/reference/templates/functions/include/) and [`includeTemplate`](https://www.chezmoi.io/reference/templates/functions/includeTemplate/).

  Note that the placement of `-` for `{{- if … -}}` and `{{- else -}}` is *required* to avoid [script exec errors](https://www.chezmoi.io/user-guide/frequently-asked-questions/troubleshooting/#chezmoi-reports-chezmoi-forkexec-tmpxxxxxxxxxxxx-exec-format-error-when-executing-a-template-script).
